internal/services/discovery/discovery.go

```diff
@@ -0,0 +1,168 @@
+
+	// GetOpenshiftClusterID discovers the OpenShift cluster ID by reading the cluster ID from the OpenShift cluster version resource.
+	GetOpenshiftClusterID(ctx context.Context) (*string, error)
+
+	// GetOpenshiftClusterName discovers the OpenShift cluster name by reading the cluster name from the OpenShift master machine resource.
+	GetOpenshiftClusterName(ctx context.Context) (*string, error)
+}
+
+var _ Service = (*ServiceImpl)(nil)
+
+type ServiceImpl struct {
+	clientset kubernetes.Interface
+	dyno      dynamic.Interface
+
+	kubeSystemNamespace   *v1.Namespace
+	kubeSystemNamespaceMu *sync.Mutex
+}
+
+func New(clientset kubernetes.Interface, dyno dynamic.Interface) *ServiceImpl {
+	return &ServiceImpl{
+		clientset:             clientset,
+		dyno:                  dyno,
+		kubeSystemNamespaceMu: &sync.Mutex{},
+	}
+}
+
+func (s *ServiceImpl) GetCSPAndRegion(ctx context.Context) (csp, region *string, reterr error) {
+	return s.getCSPAndRegion(ctx, "")
+}
+
+func (s *ServiceImpl) GetClusterID(ctx context.Context) (*uuid.UUID, error) {
+	ns, err := s.getKubeSystemNamespace(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	clusterID, err := uuid.Parse(string(ns.UID))
+	if err != nil {
+		return nil, fmt.Errorf("parsing namespace %q uid: %w", metav1.NamespaceSystem, err)
+	}
+
+	return &clusterID, nil
+}
+
+func (s *ServiceImpl) getKubeSystemNamespace(ctx context.Context) (*v1.Namespace, error) {
+	s.kubeSystemNamespaceMu.Lock()
+	defer s.kubeSystemNamespaceMu.Unlock()
+
+	if s.kubeSystemNamespace != nil {
+		return s.kubeSystemNamespace, nil
+	}
+
+	ns, err := s.clientset.CoreV1().Namespaces().Get(ctx, metav1.NamespaceSystem, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("getting namespace %q: %w", metav1.NamespaceSystem, err)
+	}
+
+	s.kubeSystemNamespace = ns
+
+	return ns, nil
+}
+
+func (s *ServiceImpl) getCSPAndRegion(ctx context.Context, next string) (csp, region *string, reterr error) {
+	nodes, err := s.clientset.CoreV1().Nodes().List(ctx, metav1.ListOptions{Limit: 10, Continue: next})
+	if err != nil {
+		return nil, nil, fmt.Errorf("listing nodes: %w", err)
+	}
+
+	for i, n := range nodes.Items {
+		ready := false
+
+		for _, cond := range n.Status.Conditions {
+			if cond.Type == v1.NodeReady && cond.Status == v1.ConditionTrue {
+				ready = true
+				break
+			}
+		}
+
+		if !ready {
+			continue
+		}
+
+		nodeCSP, ok := getCSP(&nodes.Items[i])
+		if ok {
+			csp = &nodeCSP
+		}
+
+		nodeRegion, ok := getRegion(&nodes.Items[i])
+		if ok {
+			region = &nodeRegion
+		}
+
+		if csp != nil && region != nil {
+			return csp, region, nil
+		}
+	}
+
+	if nodes.Continue != "" {
+		return s.getCSPAndRegion(ctx, nodes.Continue)
+	}
+
+	var properties []string
+	if csp == nil {
+		properties = append(properties, "csp")
+	}
+	if region == nil {
+		properties = append(properties, "region")
+	}
+
+	return nil, nil, fmt.Errorf("failed discovering properties: %s", strings.Join(properties, ", "))
```

Comment:
change csp/region to non-pointers, change checks from `==nil` to `==""`, and here just print values.

```suggestion
	return nil, nil, fmt.Errorf("failed discovering cps and region: csp=%q, region=%q", csp, region)
```

Reply:
ah, sorry, just saw this is a moved code. kindly asking to boyscout then:)

Reply:
Redone the return values to be point'less :smirk: